addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/thirdgeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,493 @@
+        if (refreshScheduler != null) {
+            refreshScheduler.cancel(true);
+			refreshScheduler = null;
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(ThirdGenerationConfiguration.class);
+        // temporary value while initializing
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Start the authentication
+        scheduler.schedule(this::authenticate, 1, TimeUnit.SECONDS);
+
+        // Start the update scheduler as configured
+        refreshScheduler = scheduler.scheduleWithFixedDelay(this::updateChannelValues, 10,
+                config.refreshInternalInSeconds, TimeUnit.SECONDS);
+    }
+
+    /**
+     * The API supports the resolution of multiple values at a time
+     *
+     * Therefore this methods builds one request to gather all information for the current inverter.
+     * The list contains all channels as defined in {@link ThirdGenerationMappingInverterToChannel} for the
+     * current inverter
+     *
+     */
+    void updateChannelValues() {
+        Map<String, List<ThirdGenerationChannelMappingToWebApi>> channelList = ThirdGenerationMappingInverterToChannel
+                .getModuleToChannelsMappingForInverter(inverterType);
+        JsonArray updateMessageJsonArray = getUpdateChannelMessage(channelList);
+
+        // Send the API request to get values for all channels
+        ContentResponse updateMessageContentResponse;
+        try {
+            updateMessageContentResponse = ThirdGenerationHttpHelper.executeHttpPost(httpClient, config.url,
+                    PROCESSDATA, updateMessageJsonArray, sessionId);
+            if (updateMessageContentResponse.getStatus() == 404) {
+                // Module not found
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        COMMUNICATION_ERROR_INCOMPATIBLE_DEVICE);
+                return;
+            }
+            if (updateMessageContentResponse.getStatus() == 503) {
+                // Communication error (e.g. during initial boot of the SCB)
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        COMMUNICATION_ERROR_HTTP);
+                return;
+            }
+            if (updateMessageContentResponse.getStatus() == 401) {
+                // session not valid (timed out?)
+                logger.info("Session expired");
+                authenticate();
+                // Retry
+                updateMessageContentResponse = ThirdGenerationHttpHelper.executeHttpPost(httpClient, config.url,
+                        PROCESSDATA, updateMessageJsonArray, sessionId);
+            }
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            // Communication problem
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, COMMUNICATION_ERROR_HTTP);
```

Comment:
`InterruptedException` should be handled differently. Just call `Thread.currentThread().interrupt();`

The binding is set here off line, but is only set online when authenticated again. Is that always working correctly? I would expect a check if data is handled it should check if the status isn't online and if that is the case set it to online.

Reply:
Added the Interrupt() call (also other lines where this exception is handled)

Also setting the device online at the end of the method.
Giving it a longer thought: In case there is a network communication (packet loss), this could have been a issue. Thanks for pointing that out!